pkg/kubelet/volumemanager/populator/desired_state_of_world_populator.go

```diff
@@ -188,15 +177,7 @@ func (dswp *desiredStateOfWorldPopulator) findAndAddNewPods() {
 	}
 
 	for _, pod := range dswp.podManager.GetPods() {
-		// Keep consistency of adding pod during reconstruction
-		if dswp.hasAddedPods && dswp.podStateProvider.ShouldPodContainersBeTerminating(pod.UID) {
-			// Do not (re)add volumes for pods that can't also be starting containers
-			continue
-		}
-
-		if !dswp.hasAddedPods && dswp.podStateProvider.ShouldPodRuntimeBeRemoved(pod.UID) {
-			// When kubelet restarts, we need to add pods to dsw if there is a possibility
-			// that the container may still be running
+		if dswp.podStateProvider.ShouldPodRuntimeBeRemoved(pod.UID) {
```

Comment:
These changes have these possible broad performance effects:
1) we may also start to mount volumes unnecessarily if a Pod is created and terminated quickly.
2) All volume types, including non-remountable types, will be reprocessed every iteration.

I think to de-risk the impact, it would be better to make this fix more targeted to:
1) volumes that have already been mounted
2) volume types that support refreshing

There is probably some parts of https://github.com/kubernetes/kubernetes/pull/107824 that could be used.

There are also implications and possible dependencies on the new volume reconstruction as well.

Reply:
Thanks, @msau42! I will take a look in the coming days.